root/static/scripts/edit/utility/linkPhrase.js

```diff
@@ -1,107 +1,194 @@
-    result[info.attribute.name] = id;
-}
+  // remove {foo} {bar} junk, unless it's for a required attribute.
+  return clean(phrase.replace(attributeRegex, function (match, name, alt) {
+    const id = idsByName[name];
+    const attr = id && linkType.attributes ? linkType.attributes[id] : null;
 
-exports.clean = _.memoize(function (linkTypeID, backward) {
-    var linkType = linkTypeInfo.byId[linkTypeID];
-    var idsByName = _.transform(linkType.attributes, mapNameToID);
+    if (attr && !attr.min) {
+      return (alt ? alt.split('|')[1] : '') || '';
+    }
 
-    // remove {foo} {bar} junk, unless it's for a required attribute.
-    var phrase = backward ? linkType.reversePhrase : linkType.phrase;
+    return match;
+  }));
+}, (a, b) => String(a.id) + b);
+
+const emptyResult = Object.freeze(['', '']);
+
+type CachedResult = {|
+  attributeValues: ?{__proto__: null, [string]: Array<string>},
+  phraseAndExtraAttributes: {__proto__: null, [string]: [string, string]},
+|};
+
+const resultCache = new WeakMap<RelationshipT, CachedResult>();
+
+function _getResultCache(relationship: RelationshipT) {
+  let result = resultCache.get(relationship);
+  if (!result) {
+    result = {
+      attributeValues: null,
+      phraseAndExtraAttributes: Object.create(null),
+    };
+    resultCache.set(relationship, result);
+  }
+  return result;
+}
 
-    return clean(phrase.replace(attributeRegex, function (match, name, alt) {
-        var id = idsByName[name];
+function getAttributeLName(type: LinkAttrTypeT) {
+  if (type.rootID === INSTRUMENT_ROOT_ID) {
+    return l_instruments(type.name);
+  }
+  return l_relationships(type.name);
+}
 
-        if (id !== undefined && linkType.attributes[id].min < 1) {
-            return (alt ? alt.split('|')[1] : '') || '';
-        }
+function _setAttributeValues(relationship: RelationshipT, cache: CachedResult) {
+  const attributeValues = Object.create(null);
+  cache.attributeValues = attributeValues;
+
+  const attributes = relationship.attributes;
+  for (let i = 0; i < attributes.length; i++) {
+    const attribute = attributes[i];
+    const type = linkedEntities.link_attribute_type[attribute.type.id];
+    const typeName = getAttributeLName(type);
+    let value = typeName;
+
+    if (type.freeText) {
+      value = clean(attribute.text_value);
+      if (value) {
+        value = l('{attribute}: {value}', {attribute: typeName, value: value});
+      }
+    }
 
-        return match;
-    }));
-}, (a, b) => a + String(b));
+    if (type.creditable) {
+      const credit = clean(attribute.credited_as);
+      if (credit) {
+        value = l('{attribute} [{credited_as}]', {
+          attribute: typeName,
+          credited_as: credit,
+        });
+      }
+    }
 
-exports.interpolate = function (linkType, attributes) {
-    if (!linkType) {
-        return ['', '', ''];
+    if (value) {
+      const rootName = linkedEntities.link_attribute_type[type.rootID].name;
```

Comment:
While trying this, I'm getting TypeError: Cannot read property 'name' of undefined here, because type.rootID is 0:

{ id: 788,
  entityType: 'link_attribute_type',
  description: null,
  rootID: 0,
  gid: 'a59c5830-5ec7-38fe-9a21-c7ea54f6650a',
  childOrder: null,
  creditable: false,
  freeText: true,
  name: 'number',
  parentID: null }

Reply:
I think you might have to clear redis before testing this; can you running `redis-cli flushdb` and see if that works?

Reply:
Yeah, that works fine actually, sorry!